src/handler/core.rs

```diff
@@ -159,17 +292,61 @@ where
     {
         let required_update_kinds_set = self.description().merge_branch(next.description());
 
-        from_fn_with_description(required_update_kinds_set, move |event, cont| {
-            let this = self.clone();
-            let next = next.clone();
-
-            this.execute(event, |event| async move {
-                match next.dispatch(event).await {
-                    ControlFlow::Continue(event) => cont(event).await,
-                    done => done,
+        let new_sig = match (&self.data.sig, &next.data.sig) {
+            (_, HandlerSignature::Entry) => {
+                panic!("Ill-typed handler branch: the second handler cannot be an entry")
+            }
+            (HandlerSignature::Entry, HandlerSignature::Other { .. }) => next.data.sig.clone(),
+            (
+                HandlerSignature::Other {
+                    input_types: self_input_types,
+                    output_types: self_output_types,
+                    obligations: self_obligations,
+                },
+                HandlerSignature::Other {
+                    input_types: next_input_types,
+                    output_types: next_output_types,
+                    obligations: next_obligations,
+                },
+            ) => {
+                HandlerSignature::Other {
+                    // Since any of the two handlers can end up being executed, take the union of
+                    // the input types of both handlers.
+                    input_types: self_input_types.union(next_input_types).cloned().collect(),
+
+                    // Since any of the two handlers can end up being executed, take the
+                    // intersection of the output types of both handlers.
+                    output_types: next_output_types
+                        .intersection(self_output_types)
+                        .cloned()
+                        .collect(),
```

Comment:
that doesn't sound right? I `.branch` is only supposed to output things from self, not next

Reply:
Doesn't the comment clarify this situation? Any of the two handlers can be executed, so we take the intersection of their output types as an approximation.